internal/callbacker/dispatcher.go

```diff
@@ -87,22 +114,75 @@ func (m *sendManager) Add(token string, dto *Callback) {
 }
 
 func (m *sendManager) GracefulStop() {
-	m.wg.Wait() // wait for all accepted callbacks to be consumed
+	m.stop <- struct{}{} // signal the `run` goroutine to stop processing
+	m.wg.Wait()          // wait for all accepted callbacks to be consumed
+
 	close(m.ch) // signal the `run` goroutine to exit
+
+	<-m.stop // wait for the `run` goroutine to exit
+	close(m.stop)
```

Comment:
Why not use a cancel context here? That would be more consistent to how we stop go routines in other services e.g. here
* https://github.com/bitcoin-sv/arc/blob/d357d5eb57cf627037110bc3a1ce26126b22432b/internal/metamorph/processor.go#L222
* https://github.com/bitcoin-sv/arc/blob/d357d5eb57cf627037110bc3a1ce26126b22432b/internal/metamorph/processor.go#L393

Reply:
I use the `stop` channel in both directions:
1. to send information to stop processing
2. to receive information that the goroutine has finished

How can I achieve this with context cancellation?

Reply:
I think it could be done like this with a second wait group. But if not no worries. We could do it later if needed.

```

func (m *sendManager) GracefulStop() {
	m.cancelAll()

	m.wg.Wait() // wait for all accepted callbacks to be consumed

	m.wgRun.Wait()

	close(m.ch) // signal the `run` goroutine to exit

}

func (m *sendManager) run() {

	ctx, cancelAll := context.WithCancel(context.Background())

	m.ctx = ctx
	m.cancelAll = cancelAll

	go func() {
		var danglingCallbacks []*store.CallbackData

		defer func() {
			m.wgRun.Done()
		}()

	handleCallbacks:
		for {
			select {
			case callback, ok := <-m.ch:
				if !ok {
					break handleCallbacks
				}

				if m.stopping {
					// add callback to save
					danglingCallbacks = append(danglingCallbacks, toStoreDto(m.url, callback))
				} else {
					m.c.Send(m.url, callback.token, callback.data)
					time.Sleep(m.sleep)
				}

				m.wg.Done()

			case <-m.ctx.Done():
				m.stopping = true
			}
		}

		_ = m.s.SetMany(context.Background(), danglingCallbacks)
		m.stop <- struct{}{}
	}()
}
```

Reply:
I'll try to change it in next PR